### PR TITLE
#479: add math Shape2D destructure extensions and MapRenderer use extension

### PR DESCRIPTION
**Pull request by Quillraven**
PR for #479

Unfortunately, the `beginRender()` and `endRender()` methods of the `OrthogonalTiledMapRenderer` are protected, which means we cannot call them directly.
I never needed a custom renderer or had to override those methods. Also, they seem to never change in recent LibGDX versions. That's why I think "copy&pasting" their content is fine in this case. However, I still added a remark in the ReadMe.md that this can be a potential issue.

If you have a problem with it then feel free to remove those extensions :)

### CODE REVIEW

File: math/README.md [hunk 1]
```diff
@@ -36,6 +36,12 @@ used to quickly determine which velocity or force is greater. Note that length *
 much faster to calculate and yields the same results in most cases.
 - `dot` infix function allows to calculate the dot product of 2 vectors.
 - `x` infix function allows to calculate the cross product of 2 vectors.
+- `Rectangle` and `Ellipse` instances can be destructed to four float variables in one step with
+  `val (x, y, w, h) = rectangle/ellipse` syntax thanks to `component1()`, `component2()`, `component3()` and `component4()` operator methods.
+- `Circle` instances can be destructed to three float variables in one step with
+  `val (x, y, radius) = circle` syntax thanks to `component1()`, `component2()` and `component3()` operator methods.
+- `Polygon` and `Polyline` instances can be destructed to two float variables in one step with
+  `val (x, y) = polygon/polyline` syntax thanks to `component1()` and `component2()` operator methods.
```
Comment (czyzby) [inline math/README.md:44]: Please move those to a separate section, this one is related to the extensions for `Vector2`.

File: tiled/src/main/kotlin/ktx/tiled/batchTiledMapRenderer.kt [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+ * Automatically calls [BatchTiledMapRenderer.beginRender] and [BatchTiledMapRenderer.endRender].
+ * @param camera A camera to set on the renderer before [BatchTiledMapRenderer.beginRender].
+ * @param block inlined. Executed after [BatchTiledMapRenderer.beginRender] and before [BatchTiledMapRenderer.endRender].
+ */
+inline fun <T : BatchTiledMapRenderer> T.use(
+  camera: OrthographicCamera,
+  block: (T) -> Unit
+) {
+  this.setView(camera)
+  AnimatedTiledMapTile.updateAnimationBaseTime()
```
Comment (czyzby) [inline tiled/src/main/kotlin/ktx/tiled/batchTiledMapRenderer.kt:19]: I'm wondering whether this update should be locked under a boolean flag such as `updateAnimation`. Could default to `true` if this is always expected to be called before using the renderer.

**Comment by czyzby** [comment]
> Unfortunately, the `beginRender()` and `endRender()` methods of the `OrthogonalTiledMapRenderer` are protected, which means we cannot call them directly.

There's a cheesy way to get around this. Create a separate Kotlin file, set the package to match `OrthogonalTiledMapRenderer` (with a warning suppression that is doesn't match the folder structure :wink:), make utility functions or extension methods that call the `beginRender()` and `endRender()`, and finally use those in the `use` extension. This would call any overridden methods.

File: tiled/src/main/kotlin/ktx/tiled/batchTiledMapRenderer.kt (hunk 2 above)
Comment (Quillraven) [reply tiled/src/main/kotlin/ktx/tiled/batchTiledMapRenderer.kt:19]: It is called per default in LibGDX and from my understanding it is linked to the animated tiles that you can create in Tiled map editor. I personally find it very weird that this is included inside a `beginRender` method but I guess they expected that people will just call `render` most of the time which renders the entire map and includes one update of the animation timer, which does make sense.

However, imo, most games will split between background and foreground layer rendering which means you need to take care of `batch.begin`, `batch.end` yourself and at least I fell into the trap once that I forgot to call this static AnimatedTileMapTile update method :D

Anyway, to summarize, imo it does not hurt to have it as a parameter with default value true. I will include it :)

Comment (Quillraven) [reply tiled/src/main/kotlin/ktx/tiled/batchTiledMapRenderer.kt:19]: edit: nevermind; if we make the cheesy solution of yours then I anyway cannot control anymore if it gets called or not because it is part of `beginRender` of the map renderer implementation ;)

File: math/README.md (hunk 1 above)
Comment (Quillraven) [reply math/README.md]: I moved it to the end of the file

**Comment by Quillraven** [comment]
> > Unfortunately, the `beginRender()` and `endRender()` methods of the `OrthogonalTiledMapRenderer` are protected, which means we cannot call them directly.
> 
> There's a cheesy way to get around this. Create a separate Kotlin file, set the package to match `OrthogonalTiledMapRenderer` (with a warning suppression that is doesn't match the folder structure 😉), make utility functions or extension methods that call the `beginRender()` and `endRender()`, and finally use those in the `use` extension. This would call any overridden methods.

I did that trick now. Did not know about it and feels weird that this is even possible :) But a nice solution imo for this scenario.

File: tiled/README.md [hunk 3]
```diff
@@ -83,6 +83,17 @@ a certain function on them.
 
 `isEmpty` and `isNotEmpty` extension method to check if the specific collection is empty or not.
 
+### `BatchTiledMapRenderer`
+
+`use` extension method to call `beginRender()` and `endRender()` automatically before
+your render logic.
+
+**Careful**: Since `beginRender()` and `endRender()` are protected methods, they are
```
Comment (czyzby) [inline tiled/README.md:91]: Is this still relevant? I think this section can be removed.

File: tiled/src/test/kotlin/ktx/tiled/TiledMapRendererTest.kt [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  @Test
+  fun `should call beginRender and endRender without setView`() {
+    val renderer = mock<BatchTiledMapRenderer>()
+
+    renderer.use {
+      verify(renderer).beginInternal()
+      assertSame(renderer, it)
+      verify(renderer, never()).endInternal()
+    }
+    verify(renderer).endInternal()
```
Comment (czyzby) [inline tiled/src/test/kotlin/ktx/tiled/TiledMapRendererTest.kt:27]: Not sure if you can verify calls on essentially static methods, you might have to change the tests package as well to verify the protected methods.

File: tiled/README.md (hunk 3 above)
Comment (Quillraven) [reply tiled/README.md:91]: No, that can be removed. I forgot that, sorry

File: tiled/src/test/kotlin/ktx/tiled/TiledMapRendererTest.kt (hunk 4 above)
Comment (Quillraven) [reply tiled/src/test/kotlin/ktx/tiled/TiledMapRendererTest.kt:27]: Why is that necessary? Isn't it sufficient to verify if the wrapper functions got called? Because the actual functions cannot be verified since they are protected. At least to my knowledge it is not advisable to verify protected/private methods.

Comment (czyzby) [reply tiled/src/test/kotlin/ktx/tiled/TiledMapRendererTest.kt:27]: I thought it might have been a limitation of Mockito, but apparently not. The tests pass, so I'm fine with verifying the extension methods.

**Comment by czyzby** [comment]
@Quillraven I missed that you've set the `master` as the target branch instead of `develop`. Can you set another PR to the `develop` branch? Sorry!
